gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -0,0 +1,897 @@
+            if (m_poFp != nullptr) {
+                VSIFCloseL(m_poFp);
+                m_poFp = nullptr;
+            }
+            return nullptr;
+        }
+
+        if (m_poFp == nullptr) {
+            CPLDebug("FlatGeobuf", "GetNextFeature: iteration start (will attempt to open file %s)", m_pszFilename);
+            m_poFp = VSIFOpenL(m_pszFilename, "rb");
+            if (m_poFp == nullptr) {
+                CPLError(CE_Failure, CPLE_AppDefined, "Failed to open file");
+                return nullptr;
+            }
+            //m_poFp = (VSILFILE*) VSICreateCachedFile ( (VSIVirtualHandle*) m_poFp);
+
+            processSpatialIndex();
+            if (m_featuresCount == 0) {
+                CPLDebug("FlatGeobuf", "GetNextFeature: no features found in spatial index search");
+                if (m_poFp != nullptr) {
+                    VSIFCloseL(m_poFp);
+                    m_poFp = nullptr;
+                }
+                return nullptr;
+            }
+        }
+
+        if (m_processedSpatialIndex && !m_ignoreSpatialFilter)
+            m_offset = m_offsetInit + m_featureOffsets[m_foundFeatureIndices[m_featuresPos]];
+        else if (m_featuresPos > 0)
+            m_offset += m_featureSize + sizeof(uoffset_t);
+
+        OGRFeature *poFeature = new OGRFeature(m_poFeatureDefn);
+
+        VSIFSeekL(m_poFp, m_offset, SEEK_SET);
+        VSIFReadL(&m_featureSize, sizeof(uoffset_t), 1, m_poFp);
+        CPL_LSBPTR32(&m_featureSize);
+        if (m_featureBufSize == 0) {
+            m_featureBufSize = std::max(1024U * 32U, m_featureSize);
+            CPLDebug("FlatGeobuf", "GetNextFeature: m_featureBufSize: %d", m_featureBufSize);
+            m_featureBuf = static_cast<GByte *>(VSIMalloc(m_featureBufSize));
+        } else if (m_featureBufSize < m_featureSize) {
+            m_featureBufSize = std::max(m_featureBufSize * 2, m_featureSize);
+            CPLDebug("FlatGeobuf", "GetNextFeature: m_featureBufSize: %d", m_featureBufSize);
+            m_featureBuf = static_cast<GByte *>(VSIRealloc(m_featureBuf, m_featureBufSize));
+        }
+        VSIFReadL(m_featureBuf, 1, m_featureSize, m_poFp);
+
+    #ifdef DEBUG
+        const uint8_t * vBuf = const_cast<const uint8_t *>(reinterpret_cast<uint8_t *>(m_featureBuf));
+        Verifier v(vBuf, m_featureSize);
+        auto ok = VerifyFeatureBuffer(v);
+        if (!ok) {
+            CPLDebug("FlatGeobuf", "VerifyFeatureBuffer says not ok");
+            CPLDebug("FlatGeobuf", "m_offset: %zu", m_offset);
+            CPLDebug("FlatGeobuf", "m_featuresPos: %zu", m_featuresPos);
+            CPLDebug("FlatGeobuf", "featureSize: %d", m_featureSize);
+        }
+    #endif
+
+        auto feature = GetRoot<Feature>(m_featureBuf);
+        auto fid = feature->fid();
+        poFeature->SetFID(fid);
+        //CPLDebug("FlatGeobuf", "GetNextFeature: fid: %zu", fid);
+        auto ogrGeometry = readGeometry(feature);
+        if (ogrGeometry == nullptr) {
```

Comment:
cannot flatgeobuf store features without geometries ?

Reply:
The format could support it (and I can't prohibit it on a schema level without uglyness) but in my experience null geometries in various formats has caused nothing but trouble so my thought was to prohibit that in implementations and spec.

Reply:
I've reconsidered and left it as allowed in the spec.